src/Bicep.Core/Registry/OciArtifactRegistry.cs

```diff
@@ -354,12 +354,14 @@ protected override void WriteArtifactContentToCache(OciArtifactReference referen
             // TODO: IsArtifactRestoreRequired assumes there must be a ModuleMain file, which isn't true for provider artifacts
             // NOTE(stephenWeatherford): That can be solved by only writing layer data files only (see below CONSIDER)
             //   and not main.json directly (https://github.com/Azure/bicep/issues/11900)
-            var moduleFileType = result switch
-            {
-                OciModuleArtifactResult => ArtifactFileType.ModuleMain,
-                OciProviderArtifactResult => ArtifactFileType.Provider,
-                _ => throw new ArgumentException($"Unexpected artifact type \"{result.GetType().Name}\".")
+            var moduleFileType = (reference.Type, result) switch {
+                (ArtifactType.Module, OciModuleArtifactResult) => ArtifactFileType.ModuleMain,
+                (ArtifactType.Module, OciProviderArtifactResult) => throw new InvalidArtifactException($"Expected a module, but retrieved a provider."),
```

Comment:
This was previously undefined behavior. I'll make sure we have a test to cover the added strictness.